awslabs/amazon-kinesis-video-streams-pic
# Release 1.2.0

*What was changed?*
- New thread API for custom stack size. New CMake flag to specify default stack size if no size is passed in. Changes are made to be backwards compatible.
- Enhanced CI. Updated runner images, refactor for easier maintainability, speed it up.
  - New CMake flag combination action with report generation.
- Code style report (apply linter) & auto-generation of patch files.
- Fix a build issue (#265)

*Why was it changed?*
- See the commits tab for more information on the individual pull requests.

*How was it changed?*
- See the commits tab for more information on the individual pull requests.

*What testing was done for the changes?*
- CI was revamped. The new CI is passing.
- Canary for WebRTC-C updated to use this branch & no build issues or performance degradation.

Additional manual tests from producer-c using this dependency:
1. Build producer-c with this branch, copied the test over and check that the new thread API is reachable.
```log
./tst/producer_test --gtest_filter="InfoProviderApiTest.TestPicThread"
Note: Google Test filter = InfoProviderApiTest.TestPicThread
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from InfoProviderApiTest
[ RUN      ] InfoProviderApiTest.TestPicThread
2024-11-07 19:46:20.726 ERROR   defaultCreateThreadWithParams(): pthread_attr_setstacksize failed with 22
[       OK ] InfoProviderApiTest.TestPicThread (1 ms)
[----------] 1 test from InfoProviderApiTest (1 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1 ms total)
[  PASSED  ] 1 test.
```
2. Check that the build fails when both conflicting arguments are set:
```
$ cmake .. -DBUILD_DEPENDENCIES=OFF -DKVS_DEFAULT_STACK_SIZE=256 -DCONSTRAINED_DEVICE=ON
...
CMake Error at dependency/libkvspic/kvspic-src/CMakeLists.txt:62 (message):
  Conflicting parameters: KVS_DEFAULT_STACK_SIZE and CONSTRAINED_DEVICE
  cannot both be set.

-- Configuring incomplete, errors occurred!
```

3. Pass in stack size only:
```
$ cmake .. -DBUILD_DEPENDENCIES=OFF -DKVS_DEFAULT_STACK_SIZE=65536
...
[100%] Completed 'libkvspic-download'
[100%] Built target libkvspic-download
-- Building with default stack size: 65536 bytes
...
```

4. Added some local tests to Producer-C to check the stack size of threads with THREAD_CREATE and the new API as well. Tried it with a few different KVS_DEFAULT_STACK_SIZE's set and with CONSTRAINED_DEVICE and the new API. Each time, stack sizes changed depending on the parameters.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
